feat(hog): memory limits

mariusandra opened the pull request:
## Problem

People can make Hog run out of memory.

## Changes

- Calculate a size for every object on the stack. (min 8 bytes)
- If this goes over 64 MB, raise an error

## How did you test this code?

Tests for both Python and TS VMs.

mariusandra committed:
feat(hog): memory limits

mariusandra committed:
Merge branch 'master' into hograils

mariusandra committed:
bump version

benjackwhite commented on hogvm/typescript/src/execute.ts:
```diff
@@ -120,22 +124,44 @@ export function exec(code: any[] | VMState, options?: ExecOptions): ExecResult {
     }
 
     for (; ip < bytecode.length; ip++) {
+        // console.log('Mem used: ', memUsed)
```
```suggestion
```

benjackwhite commented on hogvm/typescript/src/utils.ts:
```diff
@@ -79,3 +82,29 @@ export function convertHogToJS(x: any): any {
     return x
 }
+
+export function calculateCost(object: any): any {
```
Is this not going to be itself very expensive to run?

mariusandra replied:
There's always the tradeoff of safety vs performance. I think this will be quite fast. We only run it once when something goes on the stack. Objects/arrays will of course take time to loop over, but 🤷.

We will however be wrong at times, and possibly by a few times. For example If we have an object `let o := {a: {b: {c: 'str'}}}`, then when we run `o.a` in `o.a.b.c`, we add the entire `{b: {c: 'str'}}` object to the stack in addition to `o` itself that's still there. We then count its size as if it was a new object. Internally JS is for sure sharing pointers and not duplicating memory.

However given the limited scope of what we're doing here, this might just be fine.

benjackwhite replied:
If you say so. I can't say for sure other than "interesting" 😅

Also does it protect against a reference object? I don't actually know how we are doing state modifications but if I were to set a nested property to be the value of the parent does is that protected against?

mariusandra replied:
Globals can't be modified, globals when read are always copied. Local variables are just elements on the stack.

benjackwhite commented on hogvm/typescript/src/execute.ts:
```diff
@@ -2,10 +2,12 @@ import RE2 from 're2'
 
 const DEFAULT_MAX_ASYNC_STEPS = 100
 const DEFAULT_TIMEOUT_MS = 5000 // ms
+const MAX_MEMORY = 128 * 1024 * 1024 // 128 MB
```
This feels like something that should be configurable when executing. 128mb for example is potentially wayyyyyy too high for a cdp function, especially given the state will get serialised to kafka (meaning 500 in a batch * 128mb = 64Gb memory)

mariusandra replied:
Yup, but it all runs sequentially, so a batch will never take more than 128MB.

mariusandra replied:
I don't want to put this too low because of [this](https://github.com/PostHog/posthog/pull/23564#discussion_r1670623880) inaccuracy in calculation (we might throw too early), however this will be the amount that goes on Kafka in case we pause right now. 🤔

I lowered it to 64MB. We can decrease it even more, make it configurable or introduce a "fat lane". All good for me, but I'd propose to just try as is.

benjackwhite replied:
Yeah the point is more that this would already explode if we had a function that could generate 64mb of stack memory 🤔 

I guess we can merge, add some way of measuring, create some nasty functions and decide on a much much lower value when we know what a typical function looks like.

mariusandra replied:
Before this PR we were running blind, now we get `maxMemSize` in every response `VMState`. Let's track some real world values and see what we need to change. This will already protect us against runaway infinite loops.

mariusandra committed:
fixes

mariusandra committed:
fixes and 64MB

mariusandra committed:
cycles and calculateCost

benjackwhite commented on hogvm/typescript/src/__tests__/utils.test.ts:
```diff
@@ -0,0 +1,32 @@
+        ).toBe(PTR_COST * 7 + 3 + 5 + 4 + 3 + 5)
+    })
+
+    test('calculateCost with cycles', async () => {
```
One thing I would really thing is necessary here is a test that generates a really large object and then runs it (maybe asserting a certain timeout) otherwise we don't know how this function performs...

mariusandra committed:
json recursive printing

mariusandra committed:
Merge branch 'master' into hograils

mariusandra committed:
can still print duplicate objects

mariusandra committed:
bump all the things

mariusandra commented:
This might be mergeable now. Changes:
- Made sure 1) print, 2) jsonStringify, 3) calculateCost all work with recursive structures by ignoring them (`null`)
- Made sure you can still have multple printed copies of the same nested object, just not recursively nested.
- Added few more tests to calculate the cost and get memory errors
- Added `memoryLimit` as one of the options we can pass to `execSync`/`execAsync` in the NodeJS VM. We can hook that up to an env, vary it by teams or do whatever we want. It's still at the 64MB default for now.
